### PR TITLE
Fix goroutine and memory leak in watchCertAuthorities

### LINKED ISSUE #10648: Trusted Cluster goroutine leak
### Description

It looks like [watchCertAuthorities](https://github.com/gravitational/teleport/blob/8932ed4e03c808dc127fbb24fc26fc55941b1c03/lib/reversetunnel/srv.go#L1085) is not returning when trusted clusters are removed. Which subsequently prevents the underlying `CertAuthorityWatcher` to never be closed and causes a memory leak.

Both leaks can be seen in the screenshot below:
Screenshot 2022-02-25 at 10-23-13 Teleport Health Stats - Grafana

The goroutine profile here was taken after deleting the clusters:
profile001

**What happened**:

While running the 500 trusted clusters tests I noticed that after the trusted clusters were deleted there was a leak of both memory and goroutines.

**What you expected to happen**:

When the trusted clusters are removed, all resources should return to levels they were at prior to adding the clusters.

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
1. Add a bunch of trusted clusters
2. Delete them all
3. Monitor goroutine and heap metrics

### Server Details
- Teleport version (run `teleport version`): 9.0.0-beta.1

### Client Details
- Tsh version (`tsh version`): 9.0.0-beta.1

### PR DESCRIPTION
**Pull request by rosstimothy**
The CA Watcher was blocking both on writing to a channel when the watcher
was closed and on HTTP calls that had no request timeout or context passed
to cause cancellation.

All resourceWatcher implementations that had a bug which may cause them to block
on writing to a channel forever were fixed by selecting on the write and ctx.Done.

Adding context.Context to all Get/Put/Post/Delete methods on the auth HTTPClient to
force callers to propagate context. Prior all calls used context.TODO which
prevents requests from being properly cancelled.

Add context propagation to RotateCertAuthority, RotateExternalCertAuthority,
GetCertAuthority, GetCertAuthorities. This is needed to get the correct ctx
from the CertAtuhorityWatcher all the way down to the HTTPClient that makes
the call.

Companion PR  https://github.com/gravitational/teleport.e/pull/405

Closes #10648

### CODE REVIEW

**Comment by rosstimothy** [comment]
Screenshot from adding and removing 500 trusted clusters
goroutine_leak_fix

goroutine profile difference from before and after running the test
profile001

**Comment by rosstimothy** [comment]
Note to reviewers:

Sorry the changeset is so big - 90% of the diff is just propagating the `context.Context` everywhere required.

**Comment by zmb3** [comment]
Haven't had time to dig through all the changes to find the actual fix.

Can the leaks be fixed without propagating the context through, or is that what fixes it?

I ask because now that we're versioning the `api` module to line up to a major version of teleport, we should avoid breaking API changes in the module. This means the changes to propagate context should not be backported to v9 or earlier. Obviously there will be exceptions, but if we can avoid breaking things that's preferable.

**Comment by rosstimothy** [comment]
> Haven't had time to dig through all the changes to find the actual fix.
> 
> Can the leaks be fixed without propagating the context through, or is that what fixes it?
> 
> I ask because now that we're versioning the `api` module to line up to a major version of teleport, we should avoid breaking API changes in the module. This means the changes to propagate context should not be backported to v9 or earlier. Obviously there will be exceptions, but if we can avoid breaking things that's preferable.

The only way to resolve the leaks caused by hanging in the `http.Client` are to correctly propagate the context and/or setting a proper Timeout on the client. We definitely should set a `Timeout` and I was going to create an issue for it that could be resolved in a follow up PR. I was a bit hesitant to add a timeout to all of the auth client http calls so close to the release. I wasn't sure that I would be able to test all of the scenarios possible to ensure that the timeout was sufficiently large enough to not cause issues. I suppose if we made the `Timeout` quite generous it wouldn't be too much of an issue but would definitely delay cleaning up the resources.

In an effort to resolve the leaks I thought it made more sense to go with the context propagation to release the resources as soon as possible without impacting everything with the `Timeout`. That being said, this leak is already present in `v7` and `v8` as well. So if this can't be backported as is, I can revert the context propagation and just pursue the `Timeout` route instead.

**Comment by jimbishopp** [comment]
I think using context is the right call. The thing that isn't great about context is having to propagate it throughout. I realize we want to get this into the v9 release, but I'm wondering if it would be possible to use a pattern like the standard lib's http.Request and add a WithContext to the client instead of changing all these signatures.

File: lib/auth/clt_test.go [hunk 1]
```diff
@@ -77,3 +86,100 @@ func TestClient_DialTimeout(t *testing.T) {
...
+	default:
+		t.Fatal("handler never got /v2/authorities/host/rotate/external request")
+	}
+
+	select {
+	case <-sawSlow:
+		// good.
+	default:
+		t.Fatal("handler never got /slow request")
+	}
```
Comment (jimbishopp) [inline lib/auth/clt_test.go:154]: nit: Fun trick :) Closed/empty channels return the default value.
```suggestion
	close(sawRoot)
	require.True(t, <-sawRoot, "handler never got /v2/authorities/host/rotate/external request")
	
	close(sawSlow)
	require.True(t, <-sawSlow, "handler never got /slow request")
```

**Comment by rosstimothy** [comment]
> I think using context is the right call. The thing that isn't great about context is having to propagate it throughout. I realize we want to get this into the v9 release, but I'm wondering if it would be possible to use a pattern like the standard lib's http.Request and add a WithContext to the client instead of changing all these signatures.

The problem is most of the noise is generated by adding the context to the Trust and various AccessPoint interfaces.

https://github.com/gravitational/teleport/blob/8932ed4e03c808dc127fbb24fc26fc55941b1c03/lib/services/trust.go#L22
https://github.com/gravitational/teleport/blob/632d851783ac5dbd7d1ba60f4fce8c95d81041e3/lib/auth/api.go#L165

I could scale back the auth http client changes a bit by add a `Get/Put/Post/DeleteWithContext` but I'm not sure how much that would minimize the changeset.

https://github.com/gravitational/teleport/blob/master/lib/auth/clt.go#L288

File: lib/services/watcher.go [hunk 2]
```diff
@@ -729,8 +729,13 @@ func (p *databaseCollector) getResourcesAndUpdateCurrent(ctx context.Context) er
...
 	p.current = newCurrent
-	p.DatabasesC <- databases
-	return nil
+
+	select {
+	case <-ctx.Done():
+		return trace.Wrap(ctx.Err())
+	case p.DatabasesC <- databases:
+		return nil
+	}
```
Comment (jimbishopp) [inline lib/services/watcher.go:738]: Always nice to have a return at the end of a function when possible.
```suggestion
	case p.DatabasesC <- databases:
	}
	return nil
```

File: lib/services/watcher.go [hunk 3]
```diff
@@ -845,8 +857,13 @@ func (p *appCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
...
 	p.current = newCurrent
-	p.AppsC <- apps
-	return nil
+
+	select {
+	case <-ctx.Done():
+		return trace.Wrap(ctx.Err())
+	case p.AppsC <- apps:
+		return nil
+	}
```
Comment (jimbishopp) [inline lib/services/watcher.go:866]:
```suggestion
	case p.AppsC <- apps:
	}
	return nil
```

File: lib/services/watcher.go [hunk 4]
```diff
@@ -987,8 +1014,12 @@ func (c *caCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
...
 	c.lock.Unlock()
 
-	c.CertAuthorityC <- casToSlice(newHost, newUser)
-	return nil
+	select {
+	case <-ctx.Done():
+		return trace.Wrap(ctx.Err())
+	case c.CertAuthorityC <- casToSlice(newHost, newUser):
+		return nil
+	}
```
Comment (jimbishopp) [inline lib/services/watcher.go:1022]:
```suggestion
	case c.CertAuthorityC <- casToSlice(newHost, newUser):
	}
	return nil
```

**Review by jimbishopp — COMMENTED** [review]
Edited: I wish there was another way to propagate context without having to add it to method signatures.

**Comment by jimbishopp** [comment]
> I could scale back the auth http client changes a bit by add a Get/Put/Post/DeleteWithContext but I'm not sure how much that would minimize the changeset.

I'm wasn't suggesting adding new methods with context. The thought was to not need methods with context at all to simplify the signatures.  Something roughly like this:

```go
func (c *Client) WithContext(ctx context.Context) *Client {
  c.ctx = ctx
}

func (c *Client) Context() context.Context {
  if c.ctx != nil {
     return c.ctx
  }
  return context.Background()
}

func (c *Client) GetCertAuthorities(caType types.CertAuthType, ... ) {
  c.Get(c.Context(), c.Endpoint("authorities", string(caType))
}
```


> The problem is most of the noise is generated by adding the context to the Trust and various AccessPoint interfaces.

I'm not familiar enough with those interfaces to comment specifically but would hope that they could be extended with the pattern listed above. Maybe not. From the looks of those interfaces there is not much consistency with respect to context.

I'm fine with going with this change. Just wanted to suggest another way. :)

**Review by smallinsky — APPROVED** [review]
LGTM, 
Regarding Jim's proposition: 
https://github.com/gravitational/teleport/pull/10871#issuecomment-1063497533
```
func (c *Client) WithContext(ctx context.Context) *Client {
  c2 := *c
  c2.ctx = ctx
  return &c2
}
```
The drowback of this approach is that a client needs to be created each time when external client request is handled because in current implementation a client like AccessPoint the client is shared between external calls. So for that case it reasonable to propagate a context of each external client call by `APIMethod(ctx, ...) `

**Comment by rosstimothy** [comment]
> > I could scale back the auth http client changes a bit by add a Get/Put/Post/DeleteWithContext but I'm not sure how much that would minimize the changeset.
> 
> I'm wasn't suggesting adding new methods with context. The thought was to not need methods with context at all, to simplify the signatures. Something roughly like this:
> 
> ```go
> func (c *Client) WithContext(ctx context.Context) *Client {
>   c2 := *c
>   c2.ctx = ctx
>   return &c2
> }
> 
> func (c *Client) Context() context.Context {
>   if c.ctx != nil {
>      return c.ctx
>   }
>   return context.Background()
> }
> 
> func (c *Client) GetCertAuthorities(caType types.CertAuthType, ... ) {
>   c.Get(c.Context(), c.Endpoint("authorities", string(caType))
> }
> ```
> 
> > The problem is most of the noise is generated by adding the context to the Trust and various AccessPoint interfaces.
> 
> I'm not familiar enough with those interfaces to comment specifically but would hope that they could be extended with the pattern listed above. Maybe not. From the looks of those interfaces there is not much consistency with respect to context.
> 
> I'm fine with going with this change. Just wanted to suggest another way. :)

The problem with this is that most things don't get a direct handle to `auth.Client`. They either get a [auth.ClientI](https://github.com/gravitational/teleport/blob/master/lib/auth/clt.go#L1925) or the appropriate `AccessPoint` interface for their teleport component, or some smaller subset of either via another locally defined interface.

Since `auth.Client` implements `auth.ClientI` and `auth.ClientI` embeds `services.Trust` there really isn't a way to reduce the blast radius of adding a context to the `services.Trust` interface.

File: lib/auth/clt_test.go [hunk 5]
```diff
@@ -77,3 +86,100 @@ func TestClient_DialTimeout(t *testing.T) {
...
+	require.NotEmpty(t, cfg.Credentials)
+	tlsCfg, err := cfg.Credentials[0].TLSConfig()
+	require.NoError(t, err)
+	srv.TLS = tlsCfg
+
+	srv.StartTLS()
+
+	ca := newCertAuthority(t, "test", types.HostCA)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
```
[the remark this answers is not in the document]
Comment (rosstimothy) [reply lib/auth/clt_test.go]: Reduced to 1 second

**Review by jimbishopp — APPROVED** [review]
To clarify for future consideration, I was proposing that we add WithContext to ClientI (made a mistake in the hasty example above). Creating a new client on each call was a hack to demonstrate the pattern - it is certainly not necessary. The system may be too mature by now, but I hope in the future we consider alternatives to including Context in every function signature.

**Comment by zmb3** [comment]
> I hope in the future we consider alternatives to including Context in every function signature.

Why? Sure it's painful the very first time, but this is idiomatic Go. Anything that makes a network call or needs to support cancelation should take a context as the first argument.

**Pull request merged**